File: legacy-c/src/parser.c
#include "parser.h"
#include <string.h>

int parse_message(Buffer *buf, Message *msg) {
    if (!buf || !msg || buf->len < sizeof(Message)) return -1;

    msg->id = buf->data[0];
    msg->value = (buf->data[1] << 8) | buf->data[2];
    return 0;
}
// legacy C parser

File: rust-refactored/src/parser.rs
#[derive(Debug, Clone, Copy)]
pub struct Message {
    pub id: u8,
    pub value: u16,
}

pub fn parse_message(buf: &[u8]) -> Result<Message, &'static str> {
    if buf.len() < 3 {
        return Err("Buffer too short");
    }

    let id = buf[0];
    let value = u16::from_be_bytes([buf[1], buf[2]]);

    Ok(Message { id, value })
}

